[inductor] Make AOTInductor runtime interface ABI compatible

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109673
* #109606
* __->__ #109450
* #109498
* #109436

Summary: Change AOTInductor runtime interface to take vector of AtenTensorHandle as input instead of vector of aten tensor objects.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

desertfire committed:
[inductor] Make AOTInductor runtime interface ABI compatible

Summary: Change AOTInductor runtime interface to take vector of AtenTensorHandle as input instead of vector of aten tensor objects.

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Make AOTInductor runtime interface ABI compatible"


Summary: Change AOTInductor runtime interface to take vector of AtenTensorHandle as input instead of vector of aten tensor objects.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

chenyang78 commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -141,6 +141,17 @@ AOTI_TORCH_EXPORT AOTI_TORCH_NOINLINE AOTITorchError aoti_torch_empty_strided(
     int32_t device_type,
     int32_t device_index);
 
+AOTI_TORCH_EXPORT AOTI_TORCH_NOINLINE AOTITorchError aoti_torch_from_blob(
```
nit - ```aoti_torch``` seems to be too general? What about ```aoti_tensor_from_blob```?

desertfire replied:
`aoti_torch_` is the uniform prefix for all the functions provided by `shim.h`. I can make it `aoti_torch_tensor_from_blob`.

chenyang78 commented on torch/csrc/inductor/aoti_torch/tensor_converter.cpp:
```diff
@@ -0,0 +1,56 @@
+  return result;
+}
+
+std::vector<at::Tensor> steal_handles_to_tensors(
```
If we have a clear ownership, could we have a single interface instead of two, which means transferring ownership. Furthermore, ```borrow_handles_to_tensors``` seems to be a little strange. We ```move``` handles to ```std::vector<at::Tensor>```, but keep the original ```handles``` vector, which actually has nothing after move.

desertfire replied:
Sticked with borrowing ownership for now.

desertfire committed:
Update on "[inductor] Make AOTInductor runtime interface ABI compatible"


Summary: Change AOTInductor runtime interface to take vector of AtenTensorHandle as input instead of vector of aten tensor objects.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Make AOTInductor runtime interface ABI compatible"


Summary: Change AOTInductor runtime interface to take vector of AtenTensorHandle as input instead of vector of aten tensor objects.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

desertfire commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1065,21 +1066,39 @@ def write_input_output_info(
+                    self.prefix.splice(
+                        """
+                            auto tmp_input_handles = borrow_raii_tensor_handles(input_handles);
+                            auto inputs = borrow_handles_to_tensors(tmp_input_handles);
```
The reason is `raii_handles_to_raw_handles` lives in model.so but `borrow_handles_to_tensors` is from libtorch. Here we are generating code for the non-abi-compatible mode, so we can use `borrow_handles_to_tensors`, but we can't merge `borrow_handles_to_tensors` into `raii_handles_to_raw_handles`.

desertfire commented on torch/csrc/inductor/aoti_torch/tensor_converter.cpp:
```diff
@@ -0,0 +1,56 @@
+  std::vector<AtenTensorHandle> result;
+  result.reserve(tensors.size());
+  for (auto tensor : tensors) {
+    auto allocated = new at::Tensor(std::move(tensor));
```
I am going with borrow only in this version, safe to start with. If we do steal later, we can create another RAII class to make sure the stealing and returning happen in pair.

ezyang replied:
This is not a borrow lol. The result has owning handles and you MUST deallocate them when you are done.

ezyang replied:
I am not sure I entirely understand this comment, but if when you pass a pointer across the ABI, the receiver is expected to free it, of course this is part of the API and you have to consistently expect this. I also think that stealing across the call is the right thing to do (as it allows AOTInductor model.so to promptly deallocate inputs.)

ezyang replied:
Is perhaps the confusion here when you have a shared pointer, sometimes the act of incrementing the refcount is referred to as a "borrow" (I'm specifically thinking of `reinterpret_borrow` in pybind11, which lets you coerce a handle into an object and could not be implemented any other way). We should NOT use this naming convention here because we're dealing with two levels of ownership (the outer, unique exclusive ownership via new/delete, and the inner shared pointer count on at::Tensor) and, as far as the ABI is concerned, only the outer ownership matters. And in any case, when referencing CPython documentation, borrow is always used to mean "did not increment reference count).

desertfire committed:
Update on "[inductor] Make AOTInductor runtime interface ABI compatible"


Summary: Change AOTInductor runtime interface to take vector of AtenTensorHandle as input instead of vector of aten tensor objects.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Make AOTInductor runtime interface ABI compatible"


Summary: Change AOTInductor runtime interface to take vector of AtenTensorHandle as input instead of vector of aten tensor objects.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Make AOTInductor runtime interface ABI compatible"


Summary: Change AOTInductor runtime interface to take vector of AtenTensorHandle as input instead of vector of aten tensor objects.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Make AOTInductor runtime interface ABI compatible"


Summary: Change AOTInductor runtime interface to take vector of AtenTensorHandle as input instead of vector of aten tensor objects.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ezyang commented on torch/csrc/inductor/aoti_torch/tensor_converter.h:
```diff
@@ -0,0 +1,34 @@
+// into a vector of aten tensors.
+//
+// Let's start with borrowing ownership instead of stealing, correctness first
+// and then optimize for performance
```
This isn't really a borrow though, since the vector contains at::Tensor you actually have owning references, so you are creating references (and, e.g., handles could go out of scope without invalidating the return value)

ezyang commented on torch/csrc/inductor/aoti_torch/tensor_converter.h:
```diff
@@ -0,0 +1,34 @@
+// a vector of AtenTensorHandles, and then pass that into model.so.
+//
+// Let's start with borrowing ownership instead of stealing, correctness first
+// and then optimize for performance
```
This is not... really a thing? Like a borrow costs the same as a steal; it's the new reference that is expensive

ezyang commented on torch/csrc/inductor/aoti_torch/tensor_converter.h:
```diff
@@ -0,0 +1,34 @@
+
+TORCH_API at::Tensor* tensor_handle_to_tensor_pointer(AtenTensorHandle handle);
+
+TORCH_API AtenTensorHandle tensor_pointer_to_tensor_handle(at::Tensor* tensor);
```
these also need to be annotated with ownership

ezyang commented on torch/csrc/inductor/aoti_torch/tensor_converter.cpp:
```diff
@@ -0,0 +1,39 @@
+  std::vector<at::Tensor> result;
+  result.reserve(handles.size());
+  for (auto handle : handles) {
+    auto tensor = *tensor_handle_to_tensor_pointer(handle);
```
this is what triggered the refcount bump, fyi

desertfire replied:
Yes, and calling `aoti_torch_delete_tensor_object` on `handle` will dec refcount. Because I am not doing that, I call it borrow here. I think the borrow concept is on the storage ptr, not on tensor object, which is hard to do when crossing C boundary.

ezyang replied:
This comment doesn't make sense to me.

If you increment refcount and don't have a matching decrement somewhere else, that's a memory leak. If you increment refcount, and then expect the caller/callee to handle the decrement, that's a steal. Nowhere does "borrow" figure here.

ezyang commented:
Bin says review order: "Igoring some naming refactorings there, the major part was to not pass vector of at::Tensor as input and output. So maybe the reading order can be, utils.py where the vector of tensors are converted, tensor_converter.cpp where the actual conversion happens, after that  interface.cpp is where those handles are wrappered as RAIIAtenTensorHandle`, passed into model_container.cpp, and then model.cpp, lastly codgen in wrapper.py"

ezyang commented on torch/csrc/inductor/aot_runtime/model.h:
```diff
@@ -6,27 +6,40 @@
-        std::to_string(expected_size),            \
-        ", but got ",                             \
-        std::to_string(actual_size));             \
+#define ATOI_RUNTIME_CHECK(EXPR, MSG) \
```
AOTI, not ATOI

desertfire replied:
will change

ezyang commented on torch/csrc/inductor/aot_runtime/model.h:
```diff
@@ -6,27 +6,40 @@
 #include <unordered_map>
```
I need a reminder if this file is ok to use in model.so or is predictor only LOL

ezyang replied:
It used to have ATen includes but it looks like... you got rid of all of them? So it's model.so now?

desertfire replied:
Yes, it is for model.so only. I will add warnings in those model.so only headers to not include aten/c10 headers.

ezyang commented on torch/csrc/inductor/aot_runtime/model.h:
```diff
@@ -46,11 +59,11 @@ class AOTInductorModelBase {
-  ~AOTInductorModelBase() {
-    C10_CUDA_CHECK(cudaEventDestroy(run_finished_));
+  ~AOTInductorModelBase() noexcept(false) {
+    ATOI_RUNTIME_CUDA_CHECK(cudaEventDestroy(run_finished_));
```
BTW, it's better to just suppress the error here if it does error, throwing in the destructor will cause an immediate terminate which will make prod teams unhappy with you

desertfire replied:
will change

ezyang commented on torch/csrc/inductor/aot_runtime/model.h:
```diff
@@ -331,45 +348,40 @@ class AOTInductorModel : public AOTInductorModelBase<AOTInductorModel> {
-inline RAIIAtenTensorHandle create_raii_tensor_handle_for_extern(
-    AtenTensorHandle handle) {
+// TODO: will update RAIIAtenTensorHandle to unique_ptr and change this
+inline RAIIAtenTensorHandle create_raii_tensor_handle(AtenTensorHandle handle) {
```
It would be great if this had "steal" in its name, because that's what it's doing.

Also, I hate having to hand-review all of the create raii tensor handle call sites. It would be much easier to verify that everything was done correctly if you have a model.so only header that wraps all of the ABI stable aoti functions into C++ functions that do RAII handling. Then, we only need to audit that file, and all other sites are guaranteed to do things correctly by RAII.

desertfire replied:
The wrapper file suggestion is good, but I can do that a later PR.

ezyang commented on torch/_inductor/codegen/aot_runtime/interface.cpp:
```diff
@@ -54,9 +52,9 @@ AOTInductorError AOTInductorModelContainerDelete(
+    AtenTensorHandle* input_handles,
     size_t num_inputs,
-    AOTInductorTensorHandle output_handles,
+    AtenTensorHandle* output_handles,
```
What's the ownership contract here? You seem to be stealing out of them below.

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1243,7 +1246,7 @@ def codegen_dynamic_dims():
                 self.prefix.writeline(f"""constants_info_[{idx}].name = "{name}";""")
                 self.prefix.writeline(
-                    f"constants_info_[{idx}].dtype = {DTYPE_TO_ATEN[tensor.dtype]};"
+                    f"constants_info_[{idx}].dtype = static_cast<int32_t>({self.codegen_dtype(tensor.dtype)});"
```
Why is the static cast needed here 🤔

ezyang reviewed (changes requested):
I can't easily tell if the ownership was done correctly. The renamings for clarity are probably the most important thing to fix first.

desertfire committed:
Update on "[inductor] Make AOTInductor runtime interface ABI compatible"


Summary: Change AOTInductor runtime interface to take vector of AtenTensorHandle as input instead of vector of aten tensor objects.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Make AOTInductor runtime interface ABI compatible"


Summary: Change AOTInductor runtime interface to take vector of AtenTensorHandle as input instead of vector of aten tensor objects.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

desertfire commented:
> I can't easily tell if the ownership was done correctly. The renamings for clarity are probably the most important thing to fix first.

Still using our previous example in https://github.com/pytorch/pytorch/pull/109022#issuecomment-1719694841, what has this PR changed is the input/output/contant ownership, and how the temp tensor handle works has not changed. Here is the before and after comparison,

```
// Before
void AOTInductorModel::run_impl(
    std::vector<at::Tensor>& args,
    std::vector<at::Tensor>& outputs,
    cudaStream_t stream,
    ProxyExecutor* proxy_executor) {
    auto arg0_1 = create_raii_tensor_handle_for_extern(reinterpret_cast<AtenTensorHandle>(&args[0])); // This effectively borrows 
    auto arg1_1 = create_raii_tensor_handle_for_extern(reinterpret_cast<AtenTensorHandle>(&args[1])); // This effectively borrows 
    auto output_tensor_handle_0 = create_raii_tensor_handle_for_extern(reinterpret_cast<AtenTensorHandle>(&outputs[0]));
....
}

// After

// On Predictor, before calling AOTInductorModelContainerRun in model.so
// Here create_handles_from_tensors will create new tensor objects on heap and will be stolen by RAIIAtenTensorHandle
        std::vector<AtenTensorHandle> input_handles =
            torch::aot_inductor::create_handles_from_tensors(input_tensors);
        std::vector<AtenTensorHandle> output_handles =
            torch::aot_inductor::create_handles_from_tensors(output_tensors);

// In model.so
AOTInductorError AOTInductorModelContainerRun(
    AOTInductorModelContainerHandle container_handle,
    // array of raw AtenTensorHandle for input tensors, and will be stolen by
    // RAIIAtenTensorHandle
    AtenTensorHandle* input_handles,
    size_t num_inputs,
    // array of raw AtenTensorHandle for output tensors, and will be stolen by
    // RAIIAtenTensorHandle
    AtenTensorHandle* output_handles,
    size_t num_outputs,
    ...

void AOTInductorModel::run_impl(
    std::vector<RAIIAtenTensorHandle>& input_handles,
    std::vector<RAIIAtenTensorHandle>& output_handles,
    cudaStream_t stream,
    AOTIProxyExecutorHandle proxy_executor) {
    auto& inputs = input_handles;
    auto& outputs = output_handles;
    auto arg0_1 = inputs[0];
    auto arg1_1 = inputs[1];
```

I do plan to make RAIIAtenTensorHandle use unique_ptr instead of shared_ptr, but want to do that step by step.

desertfire committed:
Update on "[inductor] Make AOTInductor runtime interface ABI compatible"


Summary: Change AOTInductor runtime interface to take vector of AtenTensorHandle as input instead of vector of aten tensor objects.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Make AOTInductor runtime interface ABI compatible"


Summary: Change AOTInductor runtime interface to take vector of AtenTensorHandle as input instead of vector of aten tensor objects.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Make AOTInductor runtime interface ABI compatible"


Summary: Change AOTInductor runtime interface to take vector of AtenTensorHandle as input instead of vector of aten tensor objects.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ezyang commented:
The generated code here is incomplete since you don't have the actual RAIIAtenTensorHandle calls anywhere

ezyang commented on torch/csrc/inductor/aoti_torch/tensor_converter.h:
```diff
@@ -0,0 +1,33 @@
+// a vector of AtenTensorHandles, and then pass that into model.so. Right now we
+// create new references and return their raw pointers as a vector, and the
+// returned raw pointers will be wrapped by RAIIAtenTensorHandle in model.so.
+TORCH_API std::vector<AtenTensorHandle> create_handles_from_tensors(
```
In fact, I might even suggest you create a version of the RAII class in Predictor side, so that you can return a vector of that RAII class, and then explicitly write `.release()` at the point where the steal happens, which would make it obvious to a code reader that an ownership transfer is taking place.

However, one inconvenience is that you must create an `AtenTensorHandle*` pointer for ABI reasons. So the RAII class would have to be on the entire vector, not just the elements.

ezyang commented on test/cpp/aot_inductor/test.cpp:
```diff
@@ -78,9 +79,9 @@ TEST(AotInductorTest, BasicTest) {
   AOTI_RUNTIME_ERROR_CODE_CHECK(AOTInductorModelContainerRun(
       container_handle,
-      inputs_handle,
+      input_handles.data(),
```
steals!

ezyang replied:
It's so wonky though, cuz the vector itself doesn't get stolen, only the elements inside it

ezyang commented on torch/csrc/inductor/aot_runtime/model.h:
```diff
@@ -330,30 +351,26 @@ class AOTInductorModel : public AOTInductorModelBase<AOTInductorModel> {
 }
 
+// TODO: will update RAIIAtenTensorHandle to unique_ptr and change this
+std::vector<AtenTensorHandle> raii_handles_to_raw_handles(
```
It is helpful to know that, although this allocates a vector, it only is doing borrows on all the handles.

The repeated O(n) copies to go from RAII to not RAII is quite tiresome, especially because nothing "actually" happens here. You can avoid having to do the copy if you make RAII class for the entire vector, and then just make sure you store the contents of the vector in a canonical form, though I'm not exactly sure if you should go straight there yet.

ezyang commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1065,21 +1066,40 @@ def write_input_output_info(
+                        """
+                    )
+                else:
+                    # This looks dumb, but can avoid creating two versions of code in the AOTInductor runtime.
```
it does look dumb! lol

ezyang commented on torch/csrc/inductor/aot_runtime/interface.h:
```diff
@@ -52,9 +53,13 @@ AOTInductorError AOTInductorModelContainerDelete(
     AOTInductorModelContainerHandle container_handle,
-    AOTInductorTensorHandle input_handles,
+    // array of raw AtenTensorHandle for input tensors, and will be stolen by
+    // RAIIAtenTensorHandle
```
That it gets stolen by RAIIAtenTensorHandle is an unnecessary implementation detail that doesn't need to be in the public API contract. The public API contract is that this steals all the input handles, end of story.

ezyang commented on torch/csrc/inductor/aot_runtime/interface.h:
```diff
@@ -52,9 +53,13 @@ AOTInductorError AOTInductorModelContainerDelete(
-    AOTInductorTensorHandle output_handles,
+    // array of raw AtenTensorHandle for output tensors, and will be stolen by
+    // RAIIAtenTensorHandle
+    AtenTensorHandle* output_handles,
```
BTW, it's kind of weird to steal the output handles, since for these we definitely know they want to stay live beyond the extent of the function.

ezyang reviewed (approved):
Ownership seems to be in order.

desertfire committed:
Update on "[inductor] Make AOTInductor runtime interface ABI compatible"


Summary: Change AOTInductor runtime interface to take vector of AtenTensorHandle as input instead of vector of aten tensor objects.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

chenyang78 commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -150,6 +141,33 @@ AOTITorchError aoti_torch_empty_strided(
+    c10::IntArrayRef strides(strides_ptr, ndim);
+    c10::Device device{
+        static_cast<c10::DeviceType>(device_type),
+        static_cast<c10::DeviceIndex>(device_index)};
```
Note that this may break the cpu path as the formed device like ```cpu:0``` is not valid. We may need something similar to P832206536

desertfire commented:
Landed with an exported combined PR, https://github.com/pytorch/pytorch/pull/109678
